### PR TITLE
fix(aio): improve heading-anchor placement

### PR DESCRIPTION
**Pull request by petebacondarwin**

### CODE REVIEW

**Comment by gkalpak** [comment]
[docs_test][1] fails (it seems that some assertions need to be updated).

[1]: https://travis-ci.org/angular/angular/jobs/225901477

**Comment by petebacondarwin** [comment]
Good old tests :-)

**Comment by petebacondarwin** [comment]
Fixed the tests.

**Comment by gkalpak** [comment]
Not super-excited about having them on the right, but other than that it LGTM (as long as Travis is happy).

**Comment by petebacondarwin** [comment]
I noticed that @sjtrimble's new PR (#16327) puts a heading anchor on the far right (and doesn't make it invisible when not hovering. This may have been my accident but it looked quite nice.
So I have updated this PR to achieve the same across all headings.

(I also took this as an opportunity to fix the concern that @gkalpak had about the absolute positioning of icons by a generic CSS rule was too broad).

Let me know what you think.

**Comment by gkalpak** [comment]
I am not a huge fan of far-right, always-visible anchors (but I can live with that if we can't find a better solution).

Here's another idea:
If the cases where there is a bordered parent (that messs with the left-side anchor) are limited to a couple known ones, maybe it is worth keeping the left-side anchors (possibly keeping them visible with low opacity when not hovered to make them more discoverable) and special-cases the few situations where there are bordered parents (via CSS styles).

**Review by gkalpak — APPROVED** [review]
Implementation-wise LGTM (with the tests fixed) 👍

**Comment by petebacondarwin** [comment]
I just realised this breaks copy button positioning. At least on my phone

File: aio/src/styles/2-modules/_code.scss [hunk 1]
```diff
@@ -88,14 +88,19 @@ aio-code.headed-code {
...
   border: none;
   cursor: pointer;
   &:hover {
     color: $mediumgray;
   }
-  
+
   @media (max-width: 480px) {
     top: 0;
     bottom: 0;
```
Comment (gkalpak) [inline aio/src/styles/2-modules/_code.scss]: This should probably be `right: 0`.

**Comment by gkalpak** [comment]
I don't see any difference with this PR (vs master). Positioning on mobile is probably already broken (see https://github.com/angular/angular/pull/16336/files#r113424809).
(Unless you mean something else.)

File: aio/src/styles/2-modules/_code.scss (hunk 1 above)
Comment (petebacondarwin) [inline aio/src/styles/2-modules/_code.scss]: I agree.

**Comment by petebacondarwin** [comment]
I fixed the mobile alignment. And I also made it invisible again.

**Comment by sjtrimble** [comment]
@petebacondarwin It wasn't my intent to put the anchor on the right :) but I know this might fix some of the header indentation issues for now. The side effect of your updates is having the h1 before element (the line under the h1) extend a bit further out than planned. I will submit an update.

**Comment by petebacondarwin** [comment]
Let's check this is what we actually want before you make any more changes

**Comment by petebacondarwin** [comment]
In our meeting we agreed to move the anchor back to the left side and change the CSS so that if the heading is inside a subsection the anchor is moved further left so that from a user's perspective the anchor always appears in the left gutter of the page.

**Comment by wardbell** [comment]
@petebacondarwin I like the new _anchor-in-the-gutter_.  Looking forward to seeing it work with HTML heading tags as well as MD heading tags. Thanks!

File: aio/tools/transforms/post-process-package/processors/post-process-html.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * @dgProcessor postProcessHtml
+ *
+ * @description
+ * Use the rehype processing engine to manipulate the
+ * `renderedContent` HTML via rehype "plugins" that work with HTML ASTs (HASTs).
+ * See https://github.com/wooorm/rehype
+ *
+ * Each plugin is a factory function that will be called with the "rehype" engine as `this`.
+ * The factory should return a function that takes a HAST and returns a `boolean` or `undefined`.
+ * The HAST can be mutated by the this function.
```
Comment (gkalpak) [inline aio/tools/transforms/post-process-package/processors/post-process-html.js]: _the this_ --> _the_ or _this_

**Review by gkalpak — APPROVED** [review]

File: aio/tools/transforms/post-process-package/processors/post-process-html.js (hunk 2 above)
Comment (petebacondarwin) [inline aio/tools/transforms/post-process-package/processors/post-process-html.js]: The `this` variable

Comment (gkalpak) [inline aio/tools/transforms/post-process-package/processors/post-process-html.js]: I still think you meant "The HAST can be mutated by this function` (where "this function" is the _"function that takes a HAST and returns a `boolean` or `undefined`"_, which in turn is returned by the plugin factory function) :grin:

Comment (petebacondarwin) [inline aio/tools/transforms/post-process-package/processors/post-process-html.js]: Oh yeah :-)

**Pull request closed without merging**